File: src/test/java/dev/mazurkiewicz/quizer/question/domain/QuestionTest.java
package dev.mazurkiewicz.quizer.question.domain;

import dev.mazurkiewicz.quizer.question.application.AnswerStatus;
import dev.mazurkiewicz.quizer.question.domain.model.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

class QuestionTest {

    private Question testedQuestion;

    @BeforeEach
    void prepareAnswer() {
        QuestionId questionId = new QuestionId(1);
        QuestionContent questionContent = new QuestionContent("Test question #1");
        QuestionImage questionImage = new QuestionImage("/path/to/img.jpg");
        List<Answer> answers = List.of(
                new Answer(AnswerType.A, new AnswerContent("question #1 answer A"), AnswerStatus.INCORRECT),
                new Answer(AnswerType.B, new AnswerContent("question #1 answer B"), AnswerStatus.INCORRECT),
                new Answer(AnswerType.C, new AnswerContent("question #1 answer C"), AnswerStatus.CORRECT),
                new Answer(AnswerType.D, new AnswerContent("question #1 answer D"), AnswerStatus.INCORRECT));
        QuestionSourceExamDate questionSourceExamDate = new QuestionSourceExamDate(6, 2022);
        testedQuestion = new Question(questionId,
                questionContent,
                answers,
                questionImage,
                questionSourceExamDate);
    }

    @Test
    void shouldReturnTrueWhenSelectedAnswerIsCorrect() {
        SelectedAnswer selectedAnswer = new SelectedAnswer(AnswerType.C);
        AnswerResult expectedAnswerResult = new AnswerResult(AnswerStatus.CORRECT, AnswerType.C);

        AnswerResult answerResult = testedQuestion.checkAnswer(selectedAnswer);

        assertEquals(expectedAnswerResult, answerResult);
    }

    @Test
    void shouldReturnFalseWhenSelectedAnswerIsIncorrect() {
        SelectedAnswer selectedAnswer = new SelectedAnswer(AnswerType.B);
        AnswerResult expectedAnswerResult = new AnswerResult(AnswerStatus.INCORRECT, AnswerType.C);

        AnswerResult answerResult = testedQuestion.checkAnswer(selectedAnswer);

        assertEquals(expectedAnswerResult, answerResult);
    }

    @Test
    void shouldReturnFalseWhenSelectedAnswerIsEmpty() {
        SelectedAnswer selectedAnswer = new SelectedAnswer(AnswerType.EMPTY);
        AnswerResult expectedAnswerResult = new AnswerResult(AnswerStatus.INCORRECT, AnswerType.C);

        AnswerResult answerResult = testedQuestion.checkAnswer(selectedAnswer);

        assertEquals(expectedAnswerResult, answerResult);
    }

}

File: src/test/kotlin/io/mazurkiewicz/quizer/question/domain/model/QuestionTest.kt
package io.mazurkiewicz.quizer.question.domain.model

import io.mazurkiewicz.quizer.question.infrastructure.QuestionInvalidAnswersStateException
import io.mazurkiewicz.quizer.question.infrastructure.QuestionWithoutCorrectAnswerException
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import org.junit.jupiter.api.assertThrows
import java.util.*

class QuestionTest {

    @Test
    fun `should return correct when selected answer is correct`() {
        //given
        val testedQuestion = prepareQuestion()
        val selectedAnswer = SelectedAnswer(AnswerType.C)

        //when
        val answerResult = testedQuestion.checkAnswer(selectedAnswer)

        //then
        assertThat(answerResult.status).isEqualTo(AnswerStatus.CORRECT)
        assertThat(answerResult.correctAnswer).isEqualTo(AnswerType.C)
    }

    @Test
    fun `should return incorrect when selected answer is incorrect`() {
        //given
        val testedQuestion = prepareQuestion()
        val selectedAnswer = SelectedAnswer(AnswerType.B)

        //when
        val answerResult = testedQuestion.checkAnswer(selectedAnswer)

        //then
        assertThat(answerResult.status).isEqualTo(AnswerStatus.INCORRECT)
        assertThat(answerResult.correctAnswer).isEqualTo(AnswerType.C)
    }

    @Test
    fun `should return incorrect when selected answer is empty`() {
        //given
        val testedQuestion = prepareQuestion()
        val selectedAnswer = SelectedAnswer(AnswerType.EMPTY)

        //when
        val answerResult = testedQuestion.checkAnswer(selectedAnswer)

        //then
        assertThat(answerResult.status).isEqualTo(AnswerStatus.INCORRECT)
        assertThat(answerResult.correctAnswer).isEqualTo(AnswerType.C)
    }

    @Test
    fun `should not throwing exception when answers are valid`() {
        //given
        val question = Question(
            QuestionTemplate(UUID.randomUUID()),
            QuestionAuthor(UUID.randomUUID()),
            QuestionContent("Test question"),
            listOf(
                Answer(AnswerType.A, AnswerContent("answer A"), AnswerStatus.CORRECT),
                Answer(AnswerType.B, AnswerContent("answer B"), AnswerStatus.INCORRECT),
                Answer(AnswerType.C, AnswerContent("answer C"), AnswerStatus.INCORRECT),
                Answer(AnswerType.D, AnswerContent("answer D"), AnswerStatus.INCORRECT)
            )
        )

        //expect
        assertDoesNotThrow { question.validateAnswers() }
    }

    @Test
    fun `should throw exception when answer type occurs 2 times`() {
        //given
        val question = Question(
            QuestionTemplate(UUID.randomUUID()),
            QuestionAuthor(UUID.randomUUID()),
            QuestionContent("Test question"),
            listOf(
                Answer(AnswerType.A, AnswerContent("answer A"), AnswerStatus.CORRECT),
                Answer(AnswerType.A, AnswerContent("answer A"), AnswerStatus.INCORRECT),
                Answer(AnswerType.C, AnswerContent("answer C"), AnswerStatus.INCORRECT),
                Answer(AnswerType.D, AnswerContent("answer D"), AnswerStatus.INCORRECT)
            )
        )

        //expect
        assertThrows<QuestionInvalidAnswersStateException> { question.validateAnswers() }
    }

    @Test
    fun `should throw exception when all answers are incorrect`() {
        //given
        val question = Question(
            QuestionTemplate(UUID.randomUUID()),
            QuestionAuthor(UUID.randomUUID()),
            QuestionContent("Test question"),
            listOf(
                Answer(AnswerType.A, AnswerContent("answer A"), AnswerStatus.INCORRECT),
                Answer(AnswerType.B, AnswerContent("answer B"), AnswerStatus.INCORRECT),
                Answer(AnswerType.C, AnswerContent("answer C"), AnswerStatus.INCORRECT),
                Answer(AnswerType.D, AnswerContent("answer D"), AnswerStatus.INCORRECT)
            )
        )

        //expect
        assertThrows<QuestionWithoutCorrectAnswerException> { question.validateAnswers() }
    }

    private fun prepareQuestion(): Question {
        val answers = listOf(
            Answer(AnswerType.A, AnswerContent("question #1 answer A"), AnswerStatus.INCORRECT),
            Answer(AnswerType.B, AnswerContent("question #1 answer B"), AnswerStatus.INCORRECT),
            Answer(AnswerType.C, AnswerContent("question #1 answer C"), AnswerStatus.CORRECT),
            Answer(AnswerType.D, AnswerContent("question #1 answer D"), AnswerStatus.INCORRECT)
        )
        return Question(
            QuestionTemplate(UUID.randomUUID()),
            QuestionAuthor(UUID.randomUUID()),
            QuestionContent("Test question #1"),
            answers,
            QuestionImage("/path/to/img.jpg"),
            QuestionComment(),
            QuestionId(UUID.randomUUID())
        )
    }
}